mobile/src/main/java/org/openhab/habdroid/ui/homescreenwidget/ItemUpdateWidget.kt

```diff
@@ -0,0 +1,242 @@
+            val intent = Intent(context, ItemUpdateWidget::class.java).apply {
+                action = BackgroundTasksManager.ACTION_UPDATE_WIDGET
+                putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+            }
+
+            val pendingIntent = PendingIntent.getBroadcast(context, appWidgetId, intent, 0)
+            views.setOnClickPendingIntent(R.id.outer_layout, pendingIntent)
+            views.setTextViewText(R.id.text,
+                context.getString(R.string.item_update_widget_text, data.label, data.mappedState))
+            appWidgetManager.updateAppWidget(appWidgetId, views)
+            fetchAndSetIcon(context, views, data, smallWidget, appWidgetId, appWidgetManager)
+        }
+
+        private fun fetchAndSetIcon(
+            context: Context,
+            views: RemoteViews,
+            data: ItemUpdateWidgetData,
+            smallWidget: Boolean,
+            appWidgetId: Int,
+            appWidgetManager: AppWidgetManager
+        ) = GlobalScope.launch {
+            if (data.icon.isNotEmpty()) {
+                val encodedIcon = Uri.encode(data.icon)
+                val iconFormat = context.getPrefs().getIconFormat()
+                val iconUrl = "icon/$encodedIcon?state=${data.state}&format=$iconFormat"
+
+                try {
+                    if (context.fileList().contains(getFileNameForWidget(appWidgetId))) {
+                        Log.d(TAG, "Icon exits")
+                    } else {
+                        Log.d(TAG, "Download icon")
+                        ConnectionFactory.waitForInitialization()
+                        val connection = ConnectionFactory.usableConnectionOrNull
+                        if (connection == null) {
+                            Log.d(TAG, "Got no connection")
+                            return@launch
+                        }
+                        val content = connection.httpClient.get(iconUrl).response.bytes()
+                        context.openFileOutput(getFileNameForWidget(appWidgetId), MODE_PRIVATE).use {
+                            it.write(content)
+                        }
+                    }
+                } catch (e: HttpClient.HttpException) {
+                    Log.e(TAG, "Error downloading icon for url $iconUrl", e)
+                } catch (e: IOException) {
+                    Log.e(TAG, "Error saving icon to disk", e)
+                }
+            }
+
+            setIcon(context, views, smallWidget, appWidgetId, appWidgetManager)
+        }
+
+        private fun setIcon(
+            context: Context,
+            views: RemoteViews,
+            smallWidget: Boolean,
+            appWidgetId: Int,
+            appWidgetManager: AppWidgetManager
+        ) = GlobalScope.launch {
+            val bitmap = try {
+                context.openFileInput(getFileNameForWidget(appWidgetId)).use { fileInputStream ->
+                    val inputStream = fileInputStream.readBytes().inputStream()
```

Comment:
Why are you converting an InputStream to a different InputStream here? The file stream should be totally fine by itself, shouldn't it?